### ISSUE TITLE
Unexpected Value in Preference Change Notification

### ISSUE THREAD

**Issue by tsmaeder** [eclipse-theia/theia#7480]
When a preference value is an object, the value in effect should be constructed by merging the objects from different preference scopes. So when I listen for preference changes on an object-valued preference, I would expect the the "newValue" of the preference change to be the new, merged object value. However, the "newValue" is whatever was passed to "setPreference()".

**Comment by tsmaeder** [comment]
Here's a test to illustrate the issue. 

```typescript
    it('should report a change when object preferences are set in more general scope', () => {
        prefSchema.setSchema({
            properties: {
                'test.object': {
                    type: 'object',
                    scope: 'resource'
                }
            }
        });
        let isOk = false;

        prefService.onPreferenceChanged(event => {
            if (deepEql(event.newValue, {
                first: 'zoz',
                second: 'bar'
            })) {
                isOk = true;
            }
        });

        const userProvider = getProvider(PreferenceScope.User);
        const workspaceProvider = getProvider(PreferenceScope.Workspace);
        userProvider.setPreference('test.object', { first: 'foo' });
        workspaceProvider.setPreference('test.object', { second: 'bar' });
        userProvider.setPreference('test.object', { first: 'zoz' });
        assert.ok(isOk);
    });
```

The debugger shows that a change is notified with new Value being `{ first: 'zoz' }`

**Comment by vince-fugnitto** [comment]
I'm trying to understand, do you mean you expect a combination (merge) of the user and workspace preference value? I thought that the more precise scope (workspace in this case) would be the value which takes precedence and no merging happens.

**Comment by tsmaeder** [comment]
Not for object-valued preferences, see https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration

**Comment by tsmaeder** [comment]
@vince-fugnitto thx for adding the labels.

**Comment by vince-fugnitto** [comment]
> Not for object-valued preferences, see https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration

I see, thank you for clarifying, merging can happen for preferences such as `files.exclude`.

**Comment by akosyakov** [comment]
@tsmaeder Is it about vscode compatibility or only about PreferenceService?

I think we should add compatibility tests for change events as we did for other PreferenceService APIs:
- create a VS Code extension API tests and test under different conditions
- migrate such tests to Theia integration tests (or run VS Code extension API tests against Theia)

**Comment by tsmaeder** [comment]
No, the workspace configuration API in VS Code does not report any old/new values. But _if_ we are reporting a new value in the changes from PreferenceService, I believe the "new" value should be consistent with the value we would get when asking for the value. It is not right now. It's just a matter of time until someone thinks: "I'll just use the new value" and introduces an error. Documentation clarifying the meaning of "newValue" might be enough, though.

**Comment by tsmaeder** [comment]
I just report anything bogus I find while reading through the code :-)

**Comment by colin-grant-work** [comment]
>  It's just a matter of time until someone thinks: "I'll just use the new value" and introduces an error.

I see this a lot in reviews upstream and downstream.

**Comment by colin-grant-work** [comment]
> When a preference value is an object, the value in effect should be constructed by merging the objects from different preference scopes.

Though I agree the `newValue` field should reflect the value in effect, I'm not sure about merging objects from different scopes in the way suggested here. To me, the problem in the test scenario was that an event was emitted with a value from user scope when the same preference was already overridden in workspace scope, so that the value in user scope would never take effect at all. But as for merging objects, if you had an object with two optional keys, you might want to have, say, user-scoped default `{keyA: 'foo'}`, but workspace value `{keyB: 'and not keyA in this workspace'}`, and if we merged the objects, you would always have to do something like `{keyA: null, keyB: 'this is all I really wanted'}` to make sure that you didn't accidentally pick up `keyA` from some other scope.

**Comment by tsmaeder** [comment]
I'm not sure about this issue anymore: there are two ways to go about notifying changes: we can just notify that a preference value has changed in a particular scope and let users figure out whether that changes the value _in effect_ . The other approach is to notify the _effective change_. But in that case, reporting the scope with the "changed" even does not make sense.

As for merging values: I think if we report the _effective change_, I think merging the values is the proper thing to do if we're talking about "new value". The object property name is not part of the "address" of the value. If I understand you correctly, you would report a property-based diff from the old value?
